library/core/src/contracts.rs

```diff
@@ -5,16 +5,15 @@ pub use crate::macros::builtin::{contracts_ensures as ensures, contracts_require
 /// Emitted by rustc as a desugaring of `#[ensures(PRED)] fn foo() -> R { ... [return R;] ... }`
 /// into: `fn foo() { let _check = build_check_ensures(|ret| PRED) ... [return _check(R);] ... }`
 /// (including the implicit return of the tail expression, if any).
+///
+/// This call helps with type inference for the predicate.
 #[unstable(feature = "contracts_internals", issue = "128044" /* compiler-team#759 */)]
+#[rustc_const_unstable(feature = "contracts", issue = "128044")]
 #[lang = "contract_build_check_ensures"]
 #[track_caller]
-pub fn build_check_ensures<Ret, C>(cond: C) -> impl (Fn(Ret) -> Ret) + Copy
+pub const fn build_check_ensures<Ret, C>(cond: C) -> C
 where
```

Comment:
What is the point of this function? It's just the identity function?

> This call helps with type inference for the predicate.

Surely there's a better way to do that?

Reply:
Likely. Let me give it a try

Reply:
I did not find an easy way to get rid of this call yet. I documented the reason in the design document: https://hackmd.io/Zfrs3wpAQT-PUHgsxR9PFg?view#AST-Lowering

Tbh, I think contract needs a better design. That said, I believe this PR doesn't make things worse, so I would prefer not spending too much time on this.